iceoryx_posh/include/iceoryx_posh/internal/popo/ports/client_port_roudi.hpp

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef IOX_POSH_POPO_PORTS_CLIENT_PORT_ROUDI_HPP
+#define IOX_POSH_POPO_PORTS_CLIENT_PORT_ROUDI_HPP
+
+#include "iceoryx_posh/internal/capro/capro_message.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_sender.hpp"
+#include "iceoryx_posh/internal/popo/ports/base_port.hpp"
+#include "iceoryx_posh/internal/popo/ports/client_port_data.hpp"
+#include "iceoryx_utils/cxx/optional.hpp"
+
+namespace iox
+{
+namespace popo
+{
+/// @brief The ClientPortRouDi provides the API for accessing a client port from the RouDi middleware daemon side.
+/// The client port is divided in the three parts ClientPortData, ClientPortRouDi and ClientPortUser.
+/// The ClientPortRouDi provides service discovery functionality that is based on CaPro messages. With this API the
+/// dynamic connections between clients and servers ports can be established
+class ClientPortRouDi : public BasePort
+{
+  public:
+    using MemberType_t = ClientPortData;
+
+    explicit ClientPortRouDi(cxx::not_null<MemberType_t* const> clientPortDataPtr) noexcept;
+
+    ClientPortRouDi(const ClientPortRouDi& other) = delete;
+    ClientPortRouDi& operator=(const ClientPortRouDi&) = delete;
+    ClientPortRouDi(ClientPortRouDi&& rhs) = default;
+    ClientPortRouDi& operator=(ClientPortRouDi&& rhs) = default;
+    ~ClientPortRouDi() = default;
+
+    /// @brief get an optional CaPro message that requests changes to the desired connection state of the client
+    /// @return CaPro message with desired connection state, empty optional if no state change
+    cxx::optional<capro::CaproMessage> getCaProMessage() noexcept;
```

Comment:
since it can be empty, what do you think about naming this `tryGetCaProMessage`?

Reply:
Somehow the optional expresses this already but it would be more explicit for sure. If I change this we should have this as common code style´. What do the others think? @elfenpiff @dkroenke @MatthiasKillat @ithier @mossmaurice @marthtz @FerdinandSpitzschnueffler

Reply:
Make more sense to name the function as this what it is really doing. I would personally not force to use the prefix "try" but to make it more clear. Other possibilities are also "find" which make sense when searching in a list.
Generally i vote for renaming.

Reply:
I like it!

Reply:
I also like this idea. However, isn't our intention to use optional/expected on basically most methods that return anything? If we use `try...` on kind of all methods there'd be no benefit of this prefix. Just saying. If that's what we want... and as @budrus said, lots of changes to be done.